fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,109 +4,126 @@
+        """
+        if self.enable_broadcast_buffers:
+            self.sync_buffers()
 
-        # Build the sharded optimizer
-        self.sharded_optimizer = OSS(self.module.parameters(), optim=optimizer, group=process_group, **optimizer_params)
+        # Normal FW on the base model
+        return self.base_model(*inputs, **kwargs)
 
-        # Allocate reduce buffers
-        # - Never use a bigger buffer than the number of model params
-        buffer_size = min(buffer_size, sum(p.numel() for p in self.module.parameters()))
-        self._reduce_buffers: Dict[torch.device, List[torch.Tensor]] = {}
+    def reduce(self) -> None:
+        """ .. deprecated:: 0.0.4
+            This does not need to be called, the gradient reduction is done automatically during the BW pass
+        """
+        logging.warning("This is not useful anymore, gradients have been reduced automatically with the backward pass")
 
-        # - One buffer per rank per device
-        for device, per_device in self.sharded_optimizer.per_device_params.items():
-            buffer_dtype = per_device[0][0].dtype
-            self._reduce_buffers[device] = [
-                torch.zeros(buffer_size, dtype=buffer_dtype, device=device) for _ in range(len(per_device))
+    def sync_all_params(self) -> None:
+        """
+        Sync the complete model states in between the ranks
+        """
+        with torch.no_grad():
+            work_handles = [
+                dist.broadcast(t, src=self.reference_global_rank, group=self.process_group, async_op=True)
+                for t in self.base_model.state_dict().values()
             ]
 
-        # Sanity checks
-        assert len(self.sharded_optimizer.param_to_rank) == len(
-            list(self.module.parameters())
-        ), "number of params do not match"
-        for param in self.module.parameters():
-            assert param in self.sharded_optimizer.param_to_rank, f"{param} not in the optimizer"
-
-    def __getstate__(self) -> Dict:
-        attrs = copy.copy(self.__dict__)
-        return attrs
-
-    @property
-    def optimizer(self) -> torch.optim.Optimizer:
-        return self.sharded_optimizer
-
-    def train(self, mode: bool = True) -> "ShardedDataParallel":
-        pre_mode = self.module.training
-        self.module.train(mode)
-        if self.module.training:
-            assert not self.need_reduction or pre_mode, "incorrect state transition"
-        else:
-            assert not self.need_reduction, "try to enter eval with grads unreduced"
-        return self
+        _ = list(map(lambda x: x.wait(), work_handles))
+
+    def sync_buffers(self) -> None:
+        """
+        Sync all the param buffers in between ranks (including for instance batch norm statistics).
+        """
+        with torch.no_grad():
+            for buffer in self.base_model.buffers(recurse=True):
+                dist.broadcast(buffer.data, self.reference_global_rank, self.process_group, async_op=True)
```

Comment:
any reason for not waiting for buffers?

Reply:
on the default stream, so the next operations will be queued anyway ? I can wait if that's safer, could be that it's a wrong assumption

Reply:
(fixed)